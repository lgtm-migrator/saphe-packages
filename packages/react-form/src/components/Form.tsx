import React, {
  ChangeEvent,
  Dispatch,
  FormEvent,
  ReactElement,
  SetStateAction,
  useState,
} from 'react';
import { Config as HookConfig } from '../hooks/useForm';
import useRecaptcha from '../hooks/useRecaptcha';
import { FieldTypes } from '../utils/fieldTypes';
import {
  FormState,
  getInitialFormState,
  formatFieldValue,
  getDefaultFieldValue,
} from '../utils/formHelpers';
import { Fields, FormValues, HTMLField } from '../utils/helperTypes';
import { validateField } from '../utils/validationHelpers';
import { ValidationModes } from '../utils/validationTypes';
import Field from './Field';
import FormFieldContainer from './helpers/FormFieldContainer';
import SubmitButton from './helpers/SubmitButton';

interface Props<T extends Fields> extends HookConfig<T> {
  isSubmitting: boolean;
  setIsSubmitting: Dispatch<SetStateAction<boolean>>;
}

export default function Form<T extends Fields>(props: Props<T>): ReactElement {
  const {
    name,
    fieldPack,
    validationMode = ValidationModes.AFTER_BLUR,
    fields,
    recaptcha,
    onSubmit,
    isSubmitting,
    setIsSubmitting,
  } = props;

  const [formState, setFormState] = useState<FormState>(
    getInitialFormState(fields),
  );
  const { Recaptcha, recaptchaToken } = useRecaptcha(recaptcha);

  const handleSubmit = async (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();

    let canSubmit = true;
    const errors: Record<string, string> = {};
    const touched: Record<string, boolean> = {};
    for (const [fieldName, field] of Object.entries(fields)) {
      const error = validateField(field, formState.values[fieldName]);
      errors[fieldName] = error;
      touched[fieldName] = true;
      if (error !== '') canSubmit = false;
    }

    if (!canSubmit) setFormState({ ...formState, errors, touched });
    else if (!!recaptcha && !recaptchaToken)
      console.error(recaptcha.errorMessage);
    else if (onSubmit) {
      setIsSubmitting(true);

      const submitValues: Record<string, string | number | boolean> = {};
      for (const [fieldName, field] of Object.entries(fields))
        submitValues[fieldName] = formatFieldValue(
          field,
          formState.values[fieldName] ?? '',
        );

      Promise.resolve(
        onSubmit(submitValues as FormValues<T>, { recaptchaToken }),
      );

      setIsSubmitting(false);
    }
  };

  const handleChange = (e: ChangeEvent<HTMLField>, fieldName: string) => {
    const targetValue =
      fields[fieldName]?.type === FieldTypes.CHECKBOX
        ? String((e as ChangeEvent<HTMLInputElement>).target.checked)
        : e.target.value;
    let error = formState.errors[fieldName] ?? '';

    if (
      validationMode === ValidationModes.ON_CHANGE ||
      (validationMode === ValidationModes.AFTER_BLUR &&
        formState.touched[fieldName])
    )
      error = validateField(fields[fieldName], targetValue);

    setFormState({
      ...formState,
      errors: {
        ...formState.errors,
        [fieldName]: error,
      },
      values: {
        ...formState.values,
        [fieldName]: targetValue,
      },
    });
  };

  const handleBlur = (fieldName: string) => {
    let error = formState.errors[fieldName] ?? '';
    if (
      validationMode === ValidationModes.ON_BLUR ||
      validationMode === ValidationModes.AFTER_BLUR
    )
      error = validateField(fields[fieldName], formState.values[fieldName]);

    setFormState({
      ...formState,
      errors: {
        ...formState.errors,
        [fieldName]: error,
      },
      touched: {
        ...formState.touched,
        [fieldName]: true,
      },
    });
  };

  return (
    <form onSubmit={(e) => handleSubmit(e)}>
      {Object.keys(fields).map((fieldName, index) => {
        const field = fields[fieldName];
        const fieldId = `${name}${fieldName
          .charAt(0)
          .toUpperCase()}${fieldName.slice(1)}`;
        if (field === undefined) return <div />;
        return (
          <Field
            key={index}
            id={fieldId}
            describedBy={`${fieldId}Description`}
            field={field}
            fieldPack={fieldPack}
            name={fieldName}
            description={field.description}
            error={formState.errors[fieldName] ?? ''}
            value={
              formState.values[fieldName]?.toString() ??
              getDefaultFieldValue(field).toString()
            }
            onChange={(e) => handleChange(e, fieldName)}
            onBlur={() => handleBlur(fieldName)}
          />
        );
      })}

      {recaptcha && (
        <FormFieldContainer fieldPack={fieldPack}>
          <Recaptcha />
        </FormFieldContainer>
      )}

      <SubmitButton {...{ isSubmitting, fieldPack }} />
    </form>
  );
}
