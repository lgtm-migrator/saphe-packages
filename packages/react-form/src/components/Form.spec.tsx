import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import {
  minimalNumberField,
  minimalCheckBoxField,
  minimalSelectField,
  minimalTextAreaField,
  minimalTextField,
  matchSnapshot,
} from '../utils/testHelpers';
import Form from './Form';

const fields = {
  text: minimalTextField,
  textArea: minimalTextAreaField,
  select: minimalSelectField,
  checkBox: minimalCheckBoxField,
  number: minimalNumberField,
};

describe('Form', () => {
  it('renders', () => {
    matchSnapshot(
      <Form
        isSubmitting={false}
        setIsSubmitting={jest.fn()}
        name="testForm"
        fields={fields}
      />,
    );
  });

  it('submits', async () => {
    const setIsSubmittingMock = jest.fn();
    const onSubmitMock = jest.fn();

    render(
      <Form
        isSubmitting={false}
        setIsSubmitting={setIsSubmittingMock}
        onSubmit={onSubmitMock}
        name="testForm"
        fields={fields}
      />,
    );

    userEvent.click(screen.getByText('Submit'));
    expect(setIsSubmittingMock.mock.calls.length).toBe(2);

    // TODO uncomment
    // expect(onSubmitMock.mock.calls.length).toBe(1);
  });

  it('supports change events', () => {
    const setIsSubmittingMock = jest.fn();
    const onSubmitMock = jest.fn();

    render(
      <Form
        isSubmitting={false}
        setIsSubmitting={setIsSubmittingMock}
        onSubmit={onSubmitMock}
        name="testForm"
        fields={fields}
      />,
    );

    const textField = screen.getByLabelText('Text Field') as HTMLInputElement;
    userEvent.click(textField);
    userEvent.type(textField, 'Text input');
    expect(textField.value).toBe('Text input');

    const textAreaField = screen.getByLabelText(
      'Text Area Field',
    ) as HTMLTextAreaElement;
    userEvent.click(textAreaField);
    userEvent.type(textAreaField, 'Text area input');
    expect(textAreaField.value).toBe('Text area input');

    const selectField = screen.getByLabelText(
      'Select Field',
    ) as HTMLTextAreaElement;
    userEvent.selectOptions(selectField, 'option2');
    expect(selectField.value).toBe('option2');
  });

  // it('validates', () => {
  //   let isSubmitting = false
  //   const isSubmittingMock = jest.fn((x) => isSubmitting = x);
  //   const onSubmitMock = jest.fn();

  //   render(
  //     <Form
  //       isSubmitting={isSubmitting}
  //       setIsSubmitting={isSubmittingMock}
  //       onSubmit={onSubmitMock}
  //       name="testForm"
  //       fields={{
  //         text: minimalTextField,
  //         textArea: minimalTextAreaField,
  //         select: minimalSelectField,
  //         checkBox: minimalCheckBoxField,
  //         number: minimalNumberField,
  //       }}
  //     />,
  //   );

  //   const field = screen.getByLabelText('Text Field')
  //   fireEvent.change(field, { target: {value: 'test value'} });
  //   fireEvent.blur(field);
  // })
});
