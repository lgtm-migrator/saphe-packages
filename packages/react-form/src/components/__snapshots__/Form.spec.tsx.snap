// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form renders 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
        "marginBottom": "5px",
      }
    }
  >
    <label
      htmlFor="testFormText"
    >
      Text Field
    </label>
    <input
      aria-describedby="testFormTextDescription"
      id="testFormText"
      name="text"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
      value=""
    />
    
  </div>
  <div
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
        "marginBottom": "5px",
      }
    }
  >
    <label
      htmlFor="testFormTextArea"
    >
      Text Area Field
    </label>
    <textarea
      aria-describedby="testFormTextAreaDescription"
      id="testFormTextArea"
      name="textArea"
      onBlur={[Function]}
      onChange={[Function]}
      rows={6}
      value=""
    />
    
  </div>
  <div
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
        "marginBottom": "5px",
      }
    }
  >
    <label
      htmlFor="testFormSelect"
    >
      Select Field
    </label>
    <select
      aria-describedby="testFormSelectDescription"
      id="testFormSelect"
      name="select"
      onBlur={[Function]}
      onChange={[Function]}
      value="-placeholder-"
    >
      <option
        disabled={true}
        value="-placeholder-"
      >
        
      </option>
      <option
        value="option1"
      >
        Option 1
      </option>
      <option
        value="option2"
      >
        Option 2
      </option>
    </select>
    
  </div>
  <div
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
        "marginBottom": "5px",
      }
    }
  >
    <input
      aria-describedby="testFormCheckBoxDescription"
      id="testFormCheckBox"
      name="checkBox"
      onBlur={[Function]}
      onChange={[Function]}
      type="checkbox"
      value="false"
    />
    <label
      htmlFor="testFormCheckBox"
    >
      Checkbox Field
    </label>
    
  </div>
  <div
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
        "marginBottom": "5px",
      }
    }
  >
    <label
      htmlFor="testFormNumber"
    >
      Number Field
    </label>
    <input
      aria-describedby="testFormNumberDescription"
      id="testFormNumber"
      name="number"
      onBlur={[Function]}
      onChange={[Function]}
      type="number"
      value=""
    />
    
  </div>
  <button
    disabled={false}
    type="submit"
  >
    Submit
  </button>
</form>
`;
